src/backend/renderer/utils/wayland.rs

```diff
@@ -409,18 +409,20 @@ impl SurfaceView {
 
 /// Access the buffer related states associated to this surface
 ///
-/// Calls [`compositor::with_states`] internally
-pub fn with_renderer_surface_state<F, T>(surface: &WlSurface, cb: F) -> T
+/// Calls [`compositor::with_states`] internally.
+///
+/// Returns `None`, if there never was an commit processed through `on_commit_buffer_handler`.
+pub fn with_renderer_surface_state<F, T>(surface: &WlSurface, cb: F) -> Option<T>
 where
     F: FnOnce(&mut RendererSurfaceState) -> T,
 {
     compositor::with_states(surface, |states| {
-        let mut data = states
-            .data_map
-            .get::<RendererSurfaceStateUserData>()
-            .unwrap()
-            .borrow_mut();
-        cb(&mut data)
+        if let Some(data) = states.data_map.get::<RendererSurfaceStateUserData>() {
+            let mut data = data.borrow_mut();
+            Some(cb(&mut data))
+        } else {
+            None
+        }
```

Comment:
Looks like a good candidate for Option::map
```suggestion
        states
            .data_map
            .get::<RendererSurfaceStateUserData>()
            .map(|data| cb(&mut data.borrow_mut()))
```

Or just `?`

```suggestion
        let data = states.data_map.get::<RendererSurfaceStateUserData>()?;
        Some(cb(&mut data.borrow_mut()))
```

But yeah, in general I don't care, looks good to me anyway.